File: 1]. Data Structures/01]. Array/cpp/02_bestTimeToBuy&SellStock.cpp
#include <bits/stdc++.h>
using namespace std;

void solve() {
    int n;
    cin >> n; 
    
    vector<int> prices(n);
    for(int i = 0; i < n; i++) {
        cin >> prices[i]; 
    }

  
    int min_price = prices[0];
    int max_profit = 0;

    for(int i = 0; i < n; i++) {
        min_price = min(min_price, prices[i]);
        max_profit = max(max_profit, prices[i] - min_price);
    }

    cout << max_profit << "\n";
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t = 1;
    cin >> t; 
    
    while (t--) {
        solve();
    }

    return 0;
}

File: 1]. Data Structures/01]. Array/python/02_bestTimeToBuy&SellStock.py
import sys

def solve():
    input = sys.stdin.readline
    
    t_str = input().strip()
    if not t_str:
        return
    t = int(t_str)
    
    for _ in range(t):
        n = int(input().strip())
        
        if n == 0:
            print(0)
            continue
            
        prices = list(map(int, input().strip().split()))
        
        min_price = prices[0]
        max_profit = 0
        
        for price in prices:
            if price < min_price:
                min_price = price
            elif price - min_price > max_profit:
                max_profit = price - min_price
                
        print(max_profit)

if __name__ == '__main__':
    solve()